torch/csrc/jit/pickler.cpp

```diff
@@ -0,0 +1,391 @@
+#include <torch/csrc/jit/pickler.h>
+
+namespace torch {
+namespace jit {
+
+using ::c10::IValue;
+
+const std::vector<char>& Pickler::stack() {
+  return stack_;
+}
+
+void Pickler::start() {
+  push<OpCode>(OpCode::PROTO);
+  push<uint8_t>(2);
+
+  // All attributes get pushed into a list and their indices saved in the
+  // module def
+  push<OpCode>(OpCode::EMPTY_LIST);
+  push<OpCode>(OpCode::MARK);
+}
+
+void Pickler::finish() {
+  push<OpCode>(OpCode::APPENDS);
+  push<OpCode>(OpCode::STOP);
+}
+
+void Pickler::addIValue(const IValue& ivalue) {
+  // Check if reference ivalue has been saved before
+  const void* ivalue_ptr = getPointer(ivalue);
+  if (ivalue_ptr) {
+    auto memo_entry = memo_.find(ivalue_ptr);
+    if (memo_entry != memo_.end()) {
+      // This value has already been pushed, just do a BINGET
+      pushBinGet(memo_entry->second);
+      return;
+    }
+  }
+
+  if (ivalue.isTensor()) {
+    pushTensor(ivalue);
+  } else if (ivalue.isTuple()) {
+    pushTuple(ivalue);
+  } else if (ivalue.isDouble()) {
+    pushDouble(ivalue);
+  } else if (ivalue.isInt()) {
+    // TODO: use BININT1/BININT2/LONG if possible/necessary
+    AT_ASSERT(ivalue.toInt() <= std::numeric_limits<int32_t>::max());
```

Comment:
`std::numeric_limits<int32_t>::min() - 1` passes the check but will fail to serialize correctly. This should implement LONG1 when the TODO is resolved.

Reply:
Large negative numbers will silently fail here.